// songlyrics/build/lib/duckduckgo.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.webSearch = void 0;
const request_1 = require("./request");
const VQD_REGEX = /vqd='(\d+-\d+-\d+)'/;
const SEARCH_REGEX = /DDG\.pageLayout\.load\('d',(\[.+\])\);DDG\.duckbar\.load\('images'/;
const queryString = (query) => new URLSearchParams(query).toString();
const getVQD = (query) => __awaiter(void 0, void 0, void 0, function* () {
    var _a;
    try {
        const vqdRequestUrl = new URL(`https://duckduckgo.com/?${queryString({
            q: query,
            ia: 'web',
        })}`);
        const html = yield (0, request_1.makeRequest)(vqdRequestUrl);
        return (_a = VQD_REGEX.exec(html)) === null || _a === void 0 ? void 0 : _a.at(1);
    }
    catch (err) {
        throw new Error(`Failed to get the VQD for query "${query}".`);
    }
});
// eslint-disable-next-line complexity
const webSearch = (query) => __awaiter(void 0, void 0, void 0, function* () {
    var _b, _c;
    const vqd = yield getVQD(query);
    if (!vqd) {
        throw new Error(`Failed to get the VQD for query "${query}".`);
    }
    const requestParams = new URLSearchParams({
        q: query,
        vqd,
        kl: 'wt-wt',
        l: 'en-us',
        dl: 'en',
        ct: 'US',
        sp: '1',
        df: 'a',
        ss_mkt: 'us',
        s: '0',
        bpa: '1',
        biaexp: 'b',
        msvrtexp: 'b',
        nadse: 'b',
        eclsexp: 'b',
        tjsexp: 'b',
    });
    const requestUrl = new URL(`https://links.duckduckgo.com/d.js?${requestParams.toString()}`);
    const responseString = yield (0, request_1.makeRequest)(requestUrl);
    if (/DDG.deep.is506/.test(responseString)) {
        throw new Error('A server error occurred!');
    }
    const raw = (_c = (_b = SEARCH_REGEX.exec(responseString)) === null || _b === void 0 ? void 0 : _b.at(1)) === null || _c === void 0 ? void 0 : _c.replace(/\t/g, '    ');
    if (!raw) {
        throw new Error('No results found!');
    }
    return JSON.parse(raw);
});
exports.webSearch = webSearch;
